File: desktop/src/com/inkostilation/pong/desktop/display/shapes/IShape.java
package com.inkostilation.pong.desktop.display.shapes;

import com.badlogic.gdx.Gdx;
import com.badlogic.gdx.graphics.glutils.ShapeRenderer;

import java.util.List;

public interface IShape {

    DrawRect draw(DrawRect rect, ShapeRenderer renderer);

    void addChild(IShape shape);

    List<IShape> getChildren();

    DrawRect getChildrenRect();

    boolean isReady();

    default void drawShapeTree(ShapeRenderer renderer) {
        drawShapeTree(new DrawRect(new Position(0,0), new Position(Gdx.graphics.getWidth(), Gdx.graphics.getHeight())), renderer);
    }

    default DrawRect drawShapeTree(DrawRect rect, ShapeRenderer renderer) {
        if (isReady()) {
            DrawRect tmp = draw(rect, renderer);
            DrawRect childrenPos = getChildrenRect().getNewRect(rect.getBottomLeft().getX(), rect.getBottomLeft().getY());
            for (IShape shape : getChildren()) {
                childrenPos = shape.drawShapeTree(childrenPos, renderer);
            }
            return tmp;
        }
        return rect;
    }

    class Position {
        private float x;
        private float y;

        public Position(float x, float y) {
            this.x = x;
            this.y = y;
        }

        public float getX() {
            return x;
        }

        public float getY() {
            return y;
        }

        public Position getNewPosition(float moveX, float moveY) {
            return new Position(x + moveX, y + moveY);
        }
    }

    class DrawRect {
        private Position bottomLeft;
        private Position topRight;

        public DrawRect(Position bottomLeft, Position topRight) {
            this.bottomLeft = bottomLeft;
            this.topRight = topRight;
        }

        public Position getBottomLeft() {
            return bottomLeft;
        }

        public Position getTopRight() {
            return topRight;
        }

        public DrawRect getNewRect(float moveX, float moveY) {
            return new DrawRect(bottomLeft.getNewPosition(moveX, moveY),
                    topRight.getNewPosition(moveX, moveY));
        }
    }
}

File: desktop/src/com/inkostilation/pong/desktop/display/shapes/IShape.kt
package com.inkostilation.pong.desktop.display.shapes

import com.badlogic.gdx.Gdx
import com.badlogic.gdx.graphics.glutils.ShapeRenderer

interface IShape {
    fun draw(rect: DrawRect?, renderer: ShapeRenderer): DrawRect?
    fun addChild(shape: IShape)
    val children: List<IShape>
    val childrenRect: DrawRect
    val isReady: Boolean
    fun drawShapeTree(renderer: ShapeRenderer) {
        drawShapeTree(DrawRect(Position(0, 0), Position(Gdx.graphics.width.toFloat(), Gdx.graphics.height.toFloat())), renderer)
    }

    fun drawShapeTree(rect: DrawRect?, renderer: ShapeRenderer): DrawRect? {
        if (isReady) {
            val tmp = draw(rect, renderer)
            var childrenPos = childrenRect.getNewRect(rect!!.bottomLeft!!.x, rect.bottomLeft!!.y)
            for (shape in children) {
                childrenPos = shape.drawShapeTree(childrenPos, renderer)
            }
            return tmp
        }
        return rect
    }

    class Position(val x: Float, val y: Float) {

        fun getNewPosition(moveX: Float, moveY: Float): Position {
            return Position(x + moveX, y + moveY)
        }

    }

    class DrawRect(val bottomLeft: Position?, val topRight: Position?) {

        fun getNewRect(moveX: Float, moveY: Float): DrawRect {
            return DrawRect(bottomLeft!!.getNewPosition(moveX, moveY),
                    topRight!!.getNewPosition(moveX, moveY))
        }

    }
}